account-kit/smart-contracts/src/light-account/accounts/viem-base.ts

```diff
@@ -0,0 +1,344 @@
+    TLightAccountType,
+    TLightAccountVersion
+  >
+    ? typeof entryPoint06Abi
+    : typeof entryPoint07Abi,
+  LightAccountEntryPointVersion<
+    TLightAccountType,
+    TLightAccountVersion
+  > extends "0.6.0"
+    ? "0.6"
+    : "0.7",
+  {
+    getLightAccountVersion: () => TLightAccountVersion;
+    getSigner: () => TSigner;
+    source: TLightAccountType;
+    encodeUpgradeToAndCall: (params: {
+      upgradeToAddress: Address;
+      upgradeToInitData: Hex;
+    }) => Promise<Hex>;
+  },
+  false
+>;
+
+export type ViemLightAccountBase<
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+  TLightAccountType extends LightAccountType = LightAccountType,
+  TLightAccountVersion extends
+    LightAccountVersion<TLightAccountType> = LightAccountVersion<TLightAccountType>,
+> = SmartAccount<
+  BaseLightAccountImplementation<
+    TSigner,
+    TLightAccountType,
+    TLightAccountVersion
+  >
+>;
+
+export type CreateViemLightAccountBaseParams<
+  TLightAccountType extends LightAccountType,
+  TLightAccountVersion extends
+    LightAccountVersion<TLightAccountType> = LightAccountVersion<TLightAccountType>,
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+> = {
+  client: PublicClient<TTransport, Chain>;
+  signer: TSigner;
+  abi: Abi;
+  accountAddress: Address;
+  type: TLightAccountType;
+  version: TLightAccountVersion;
+  getFactoryArgs: () => Promise<{
+    factory?: Address | undefined;
+    factoryData?: Hex | undefined;
+  }>;
+};
+
+export async function createViemLightAccountBase<
+  TLightAccountType extends LightAccountType,
+  TLightAccountVersion extends
+    LightAccountVersion<TLightAccountType> = LightAccountVersion<TLightAccountType>,
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+>({
+  client,
+  signer,
+  abi,
+  version,
+  type,
+  accountAddress,
+  getFactoryArgs,
+}: CreateViemLightAccountBaseParams<
+  TLightAccountType,
+  TLightAccountVersion,
+  TTransport,
+  TSigner
+>): Promise<
+  ViemLightAccountBase<TSigner, TLightAccountType, TLightAccountVersion>
+> {
+  const encodeUpgradeToAndCall = async ({
+    upgradeToAddress,
+    upgradeToInitData,
+  }: {
+    upgradeToAddress: Address;
+    upgradeToInitData: Hex;
+  }): Promise<Hex> => {
+    const storage = await getStorageAt(client, {
+      address: accountAddress,
+      slot: "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
+    });
+
+    if (storage == null) {
+      throw new Error(
+        "Failed to get storage slot: 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
+      );
+    }
+
+    const implementationAddresses = Object.values(
+      AccountVersionRegistry[type],
+    ).map(
+      (x) =>
+        x.addresses.overrides?.[client.chain.id]?.impl ??
+        x.addresses.default.impl,
+    );
+
+    if (
+      fromHex(storage, "number") !== 0 &&
```

Comment:
❓ random question. curious if you prefer using `fromHex` like this over just using `hexToNumber` for some reason? i prefer the latter in most cases.